psutil_agent/loop.py

```diff
@@ -0,0 +1,110 @@
+#########
+# Copyright (c) 2015 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+#
+
+import json
+import logging
+import os
+import sched
+import sys
+import time
+
+import pika
+import psutil
+from pika.exceptions import AMQPError
+
+
+def collect_metrics(rabbit_config, log_dir, psutil_config):
+    logging.basicConfig(filename=os.path.join(log_dir, 'psutil.log'))
+
+    scheduler = sched.scheduler(time.time, time.sleep)
+
+    credentials = pika.PlainCredentials(rabbit_config['broker_user'],
+                                        rabbit_config['broker_pass'])
+
+    params = pika.ConnectionParameters(credentials=credentials,
+                                       host=rabbit_config['broker_hostname'],
+                                       virtual_host=rabbit_config['vhost'],
+                                       port=rabbit_config['broker_port'])
+
+    connection = pika.BlockingConnection(params)
+    channel = connection.channel()
+    channel.exchange_declare(exchange='cloudify-monitoring',
+                             exchange_type="topic",
+                             auto_delete=True,
+                             durable=False,
+                             internal=False)
+
+    for config in psutil_config:
+        def create_scheduled_fun(method, interval, f_args, result_argument,
+                                 alias):
+            try:
+                fun = getattr(psutil, method)
+
+                def scheduled_fun():
+                    result = fun(**f_args)
```

Comment:
It's rather problematic to add a try-except just to this function, because code below it relies on the result of the invocation. Should I add a catch-all clause to the outer try-except or maybe assign None to __result__ first and then check it?